src/runtime/string.cpp

```diff
@@ -680,6 +680,56 @@ static PRIMFN(prim_hash_str) {
   RETURN(String::alloc(runtime.heap, std::move(hex)));
 }
 
+static PRIMTYPE(type_to_upper) {
+  return args.size() == 1 && args[0]->unify(Data::typeString) && out->unify(Data::typeString);
+}
+
+static PRIMFN(prim_to_upper) {
+  EXPECT(1);
+  STRING(str, 0);
+
+  const char *in = str->c_str();
+  ssize_t in_size = str->length;
+  std::string out;
+  while (in_size > 0) {
+    int32_t codepoint = 0;
+    ssize_t size = utf8proc_iterate(reinterpret_cast<const uint8_t *>(in), in_size, &codepoint);
+    in += size;
+    in_size -= size;
+    int32_t upper = utf8proc_toupper(codepoint);
+    char buf[4] = {0};
+    ssize_t bytes = utf8proc_encode_char(upper, reinterpret_cast<uint8_t *>(buf));
+    out.insert(out.end(), buf, buf + bytes);
+  }
+
+  RETURN(String::alloc(runtime.heap, std::move(out)));
+}
+
+static PRIMTYPE(type_to_lower) {
+  return args.size() == 1 && args[0]->unify(Data::typeString) && out->unify(Data::typeString);
+}
+
+static PRIMFN(prim_to_lower) {
```

Comment:
What's the difference between this new prim and the existing [`scaseNFKC`](https://github.com/sifive/wake/blob/31c5cda2c8b0864ee65e9ee29e9d95c8acc8bcb2/src/runtime/string.cpp#L520) prim, which is used by string.wake's [`unicodeLowercase`](https://github.com/sifive/wake/blob/31c5cda2c8b0864ee65e9ee29e9d95c8acc8bcb2/share/wake/lib/core/string.wake#L226-L230) function? I'm especially curious about how the existing function was basically able to delegate the entire thing to the library, without having to do any iteration on its own.

Reply:
I actually couldn't find that! I was searching for the upper case conversion which I couldn't find because it wasn't implemented. On further inspection, that primitive is kind of abusing case folding. Case folding happens in practice to lower case everything but all it says it does is that it canonicalizes the cases. There's no way to repeat that for the to upper case.

I think we should get rid of the previous primitive and I should just add `unicodeUppercase` to be consistent with naming.

Reply:
Yeah, I figured something weird was going on with it, since it wasn't obvious to me how the `UTF8PROC_CASEFOLD` flag implied lowercase. 👍 on removing it and replacing the previous implementation of `unicodeLowercase` with your new function.

Reply:
to answer directly:
the old primitive used utf8proc_map which just has a list of transformations that can be applied. One of them is called "case folding" which has unspecified semantics but maps everything into a unique form with respect to case. In practice this is lower case. Additionally you have to have pre-allocated a large enough buffer for that to be used and I see now actual assurance that such a thing is generally true for lower and upper case. The map function is allowed to return an error if the output is larger than the input (although that might never happen with upper/lower cases).

The new iterative implementation makes no such assumptions and cannot fail unless memory allocation fails.

I got rid of the old scase primitive meant for canonicalization and replaced it with the iterative version. If there was a "toupper" version of scaseNFKC I would use that instead but it doesn't appear to exist in our unicode library. In the intrest of having tolower and toupper function the same way, and as a bonus fixing those API holes we just found, think we should stick to the iterative version.

I did however switch to using the previous naming scheme